[Merged by Bors] - fix(fun_prop):  few bug fixes for fun_prop

lecopivo opened the pull request:
Bug fixes
1. Fix infinite loop triggered by `(by fun_prop : ?m)`
2. Support for multiple lambda theorems. For example `ContinuousOn` has multiple versions of composition theorem.
3. Support for special constant lambda theorem. For example `IsLinearMap R fun _ => 0` should be registered as constant lambda theorem
4. Improved error messages when trying `fun_prop` on non-fun_prop goal
5. Improved error messages when missing lambda theorems

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

lecopivo commented:
One more fix for expression involving `Expr.proj`

For example this was failing
```
example : Continuous fun ((x, y, z) : ℝ × ℝ × ℝ) ↦ y := by fun_prop
example : Continuous fun ((x, y, z) : ℝ × ℝ × ℝ) ↦ z := by fun_prop
```

YaelDillies commented:
Would you mind adding some more inline comments? The code would be easier to understand.

lecopivo commented:
Do you have a specific part in mind? I can make it more readable. It would be difficult for me to do so without knowing what to improve.

j-loreaux commented:
@alexjbest if you wouldn't mind taking a look at this at some point I would greatly appreciate it!

j-loreaux commented on Mathlib/Tactic/FunProp.lean:
````diff
@@ -9,3 +14,150 @@ import Mathlib.Tactic.FunProp.StateList
+```
+
+**Basic debugging:**
+The most common issue is that a function is missing appropriate theorem. For example
````
```suggestion
The most common issue is that a function is missing the appropriate theorem. For example
```

j-loreaux commented on Mathlib/Tactic/FunProp.lean:
````diff
@@ -9,3 +14,150 @@ import Mathlib.Tactic.FunProp.StateList
+  No theorems found for `Real.sin` in order to prove `Continuous fun x => x.sin`
+```
+this can be easily fixed by importing `Mathlib.Analysis.SpecialFunctions.Trigonometric.Basic`
+where the theorem `Real.continuous_sin` is marked with `fun_prop` attribute.
````
```suggestion
where the theorem `Real.continuous_sin` is marked with the `fun_prop` attribute.
```

j-loreaux commented on Mathlib/Tactic/FunProp.lean:
```diff
@@ -9,3 +14,150 @@ import Mathlib.Tactic.FunProp.StateList
+this can be easily fixed by importing `Mathlib.Analysis.SpecialFunctions.Trigonometric.Basic`
+where the theorem `Real.continuous_sin` is marked with `fun_prop` attribute.
+
+When the issue is not simply few missing theorems then you can turn on the option
```
```suggestion
When the issue is not simply a few missing theorems then you can turn on the option
```

j-loreaux commented on Mathlib/Tactic/FunProp.lean:
```diff
@@ -9,3 +14,150 @@ import Mathlib.Tactic.FunProp.StateList
+
+To set up `fun_prop` for a new function property you have to:
+
+1. Mark function property with `fun_prop` attribute when defining it
```
```suggestion
1. Mark the function property with `fun_prop` attribute when defining it
```

j-loreaux commented on Mathlib/Tactic/FunProp.lean:
````diff
@@ -9,3 +14,150 @@ import Mathlib.Tactic.FunProp.StateList
+  Continuous (fun x => f (g x)) := ...
+```
+The constant theorem is not absolutely necessary as for example `IsLinearMap ℝ (fun x => y)` does
+not hold but almost certainly want to mark it if it is available.
````
```suggestion
not hold but we almost certainly want to mark it if it is available.
```

j-loreaux commented on Mathlib/Tactic/FunProp.lean:
````diff
@@ -9,3 +14,150 @@ import Mathlib.Tactic.FunProp.StateList
+theorem continuous_add (f g : X → Y) (hf : Continuous f) (hg : Continuous g) :
+    Continuous (fun x => f x + g x) := ...
+```
+It is enough to provide function theorems in either form. It is manly a matter of convenience.
````
```suggestion
It is enough to provide function theorems in either form. It is mainly a matter of convenience.
```

j-loreaux commented on Mathlib/Tactic/FunProp.lean:
```diff
@@ -9,3 +14,150 @@ import Mathlib.Tactic.FunProp.StateList
+It is enough to provide function theorems in either form. It is manly a matter of convenience.
+
+
+With such basic set up you should be able to prove continuity of basic algebraic expressions.
```
```suggestion
With such a basic set up you should be able to prove continuity of basic algebraic expressions.
```

j-loreaux commented on Mathlib/Tactic/FunProp.lean:
```diff
@@ -9,3 +14,150 @@ import Mathlib.Tactic.FunProp.StateList
+
+When marking theorems it is a good idea to check that a theorem has been registered correctly.
+You can do this by turning on the `Meta.Tactic.fun_prop.attr` option. For example
+(note that the notation `f x + g x` is just a syntatic suggar for `@HAdd.hAdd X Y Y _ (f x) (g x))
```
```suggestion
(note that the notation `f x + g x` is just a syntactic sugar for `@HAdd.hAdd X Y Y _ (f x) (g x))
```

j-loreaux commented on Mathlib/Tactic/FunProp.lean:
```diff
@@ -9,3 +14,150 @@ import Mathlib.Tactic.FunProp.StateList
+  - morphism theorems
+  - transition theorems
+
+- checking theorems is correctly recognized
```
```suggestion
- checking theorems are correctly recognized
```

j-loreaux commented on Mathlib/Tactic/FunProp/Core.lean:
```diff
@@ -542,6 +573,10 @@ def constAppCase (funPropDecl : FunPropDecl) (e : Expr) (fData : FunctionData)
   if let .some r ← tryTheorems funPropDecl e fData localThms funProp then
     return r
 
+  -- log error if no global or local theoresm were found
```
```suggestion
  -- log error if no global or local theorems were found
```

j-loreaux commented on Mathlib/Tactic/FunProp/Elab.lean:
```diff
@@ -51,25 +51,31 @@ def funPropTac : Tactic
 
-      let cfg : Config := {disch := disch, constToUnfold := .ofArray namesToUnfold _}
-      let (r?, s) ← funProp goalType cfg |>.run {}
+      -- the whnf and telscope is here because the goal can be
```
```suggestion
      -- the whnf and telescope is here because the goal can be
```

j-loreaux reviewed (commented):
@lecopivo thanks for starting to work on this! All these comments are just typo corrections and should be trivially mergeable. Ping me when you have addressed Yaël's comment above too and I will take another look.

lecopivo commented on Mathlib/Tactic/FunProp.lean:
````diff
@@ -127,37 +137,213 @@ displays
+    The function theorem for `Neg.neg` and `Continuous` can be stated as:
+    ```lean
+    @[fun_prop]
+    theorem continuous_neg : Continuous (fun x => Neg.neg x) := ...
````
I'm doing it because for example for `a + b` the head of this expression is `HAdd.hAdd`  which is obvious when you write it as `HAdd.hAdd a b` and you don't even know what is the definition of "head of an expression".

Maybe should I switch to some functions that don't have this issue? Maybe `sin` and what would be a good two argument function?

j-loreaux replied:
No, that's fine, it makes sense, but perhaps a short parenthetical comment to the effect of "(the *head* of an expression is the term which appears first in the tree; for binary operations, like `a + b`, after stripping away the syntactic sugar / notation this becomes `HAdd.hAdd a b`; here we prefer the latter to make the head expression more obvious)"

lecopivo commented on Mathlib/Tactic/FunProp.lean:
```diff
@@ -127,37 +137,213 @@ displays
+    If you really care that the resulting proof term is as short as possible, it is a good idea to
+    provide both versions.
+
+    One exception to this rule is the theorem for `Prod.mk`, which has to be stated in compositional
+    form.
```
Because the theorem for `Prod.mk` is used to decompose more complicated functions into simpler. 

For example,
When you want to prove something about the function `(fun x => sin (exp x) + cos (exp x))` it get decomposed as
```
(fun x => x.1 + x.2)
∘
(fun x => (sin (exp x), cos (exp x)))
```
You should have exact theorem for `(fun x => x.1 + x.2)`. 

To deal with `fun x => (sin (exp x), cos (exp x))` you apply theorem for `Prod.mk` to decompose it into two problems, one for `fun x => sin (exp x)` and one for `fun x => cos (exp x)`.

lecopivo replied:
I added 
> This because this theorem together with *lambda theorems* is used to break expression to smaller pieces and `fun_prop`  assumes it is written in compositional form.

Does that make it a bit more clear?

lecopivo commented on Mathlib/Tactic/FunProp/Core.lean:
```diff
@@ -147,20 +146,37 @@ def tryTheoremWithHint? (e : Expr) (thmOrigin : Origin)
+  -- For example, I want to be able to solve a goal with mvars like `ContDiff ℝ ?n f` using local
+  -- hypothesis `(h : ContDiff ℝ ∞ f)` and assign `∞` to the mvar `?n`.
+  --
+  -- This could be problematic if there are two local hypothesis `(hinf : ContDiff ℝ ∞ f)` and
+  -- `(h1 : ContDiff ℝ 1 f)` and appart from solving `ContDiff ℝ ?n f` there is also a subgoal
+  -- `2 ≤ ?n`. If `fun_prop` decides to try `h1` first it would assign `1` to `?n` and then there
+  -- is no hope solving `2 ≤ 1` and it won't be able to apply `hinf` after trying `h1` as `n?` is
+  -- assigned already. Ideally `fun_prop` would roll back the `MetaM.State`. This issue did not
+  -- come up yet so I didn't bother and I'm worried about the performance impact.
```
Yeah it will probably come up. I will think about it but will not solve it right now.

lecopivo commented on Mathlib/Tactic/FunProp.lean:
````diff
@@ -127,37 +137,213 @@ displays
+    Note that without notation and coercion, the function looks like
+    `fun x => DFunLike.coe (f x) y`.
+
+    In fact, not only `DFunLike.coe` but any function coercion is treated this way. Such function
+    coercion has to be registered with `Lean.Meta.registerCoercion` with coercion type `.coeFun`.
+    ```lean
+    #eval Lean.Elab.Command.liftTermElabM do
+      Lean.Meta.registerCoercion <morphism_to_fun_coercion_name>
+        (some { numArgs := _, coercee := _, type := .coeFun })
+    ```
+    This is useful when you define your own bundled morphism and do not want to use Mathlib's
+    `FunLike` class.
````
Yes you are right. I will add an example with `MyFunLike` that will prevent the confusion.

lecopivo commented:
I think I have addressed all the comments except for the `if then else`. I disagree that every `if` should be commented, the function there is literally called `tryTheorem...`, so the `if` statement is testing if trying a theorem succeed or not.

faenuccio commented:
I have removed the `awaiting-review` label according to the decision explained here
https://leanprover.zulipchat.com/#narrow/stream/113486-announce/topic/Changes.20to.20the.20.23queue/near/450176907

joneugster commented on Mathlib/Tactic/FunProp/Types.lean:
```diff
@@ -19,14 +19,11 @@ open Lean Meta
-initialize registerTraceClass `Meta.Tactic.fun_prop.unify
-initialize registerTraceClass `Meta.Tactic.fun_prop.discharge
-initialize registerTraceClass `Meta.Tactic.fun_prop.apply
-initialize registerTraceClass `Meta.Tactic.fun_prop.unfold
+initialize registerTraceClass `Meta.Tactic.fun_prop.attr
 initialize registerTraceClass `Meta.Tactic.fun_prop.cache
+initialize registerTraceClass `Meta.Tactic.fun_prop.errors
+initialize registerTraceClass `Meta.Tactic.fun_prop.step
```
1. Should these have `(inherited := true)` so that `set_option trace.Meta.Tactic.fun_prop` enables all of them? (see [docstring of `registerTraceClass`](https://leanprover-community.github.io/mathlib4_docs/Lean/Util/Trace.html#Lean.registerTraceClass))
2. These could benefit from docstrings explaining what sort of traces they enable.

lecopivo replied:
What is the intended way to add doc to an option? 

Like this?
```
/-- bla bla bla -/
initialize registerTraceClass `Meta.Tactic.fun_prop
```

Also I will just cut the tracing options to
```
Meta.Tactic.fun_prop
Meta.Tactic.fun_prop.attr
Debug.Meta.Tactic.fun_prop
```
I don't think there is a need for more right now.

joneugster replied:
Thanks that makes it much more understandable!

Yes, if you add a docstring `/-- yada yada -/` to `initialize registerTraceClass` I think it appears on hover. However, this is completely optional for this PR!

joneugster commented on Mathlib/Tactic/FunProp/Types.lean:
```diff
@@ -66,49 +63,71 @@ def FunctionData.getFnOrigin (fData : FunctionData) : Origin :=
   thmStack : List Origin := []
-  /-- Maximum number of steps `fun_prop` can take. -/
-  maxSteps := 100000
-deriving Inhabited
+  /-- Logging mode, determines whether logged messages will be diplayes to the user immediately or
+  on request by setting `Meta.Tactic.fun_prop.errors` to true. -/
+  loggingMode : LoggingMode := .main
 
 /-- `fun_prop` state -/
 structure State where
-  /-- Simp's cache is used as the `funProp` tactic is designed to be used inside of simp and utilize
-  its cache -/
-  cache : Simp.Cache := {}
+  /-- Simp's cache is used as the `fun_prop` tactic is designed to be used inside of simp and
+  utilize its cache -/
+  cache        : Simp.Cache := {}
   /-- Count the number of steps and stop when maxSteps is reached. -/
   numSteps := 0
   /-- Log progress and failures messages that should be displayed to the user at the end. -/
-  msgLog : List String := []
+  mainMsgLog : List String := []
+  /-- Log progress and failures messages that should be displayed to the user at the end when
+  the option `Meta.Tactic.fun_prop.errors` is set to true. -/
+  secondaryMsgLog : List String := []
```
This entire approach to logging seems a bit peculiar to me. I feel like the more common ways to do this would be either:

### Option A

Use just one `msgLog` as before and no "LoggingMode". Create an option

```
register_option Meta.Tactic.fun_prop.errors : Bool := {
  defValue := false
  descr := "yada yada" }
```

and then use `if Meta.Tactic.fun_prop.errors.get (← getOptions) then` whenever you want to add a "secondary" log message. (this approach does not sort them as main/secondary, but seems much simpler)

### Option B

Just use `trace[Meta.Tactic.fun_prop.errors]` wherever you have these secondary error messages, don't bother about the sorting and have the error say something like "enable `set_option trace.Meta.Tactic.fun_prop.errors true` to see more details!" No need for the "LoggingMode" either.

### Option C
If you really think adding `LoggingMode` adds value, I'd still suggest you use `register_option` instead of `registerTraceClass`.

In any case, it would be nice to have a test in the dev-test file using `#guard_msgs` which demonstrates the expected error message, would help understanding and reviewing..

lecopivo replied:
The secondary errors are already reported with `set_option trace.Meta.Tactic.fun_prop` so I'm going to completely ditch the main/secondary dichotomy. 

I'm also going to include few tests with `#guard_msgs` to the dev file as you suggest.

joneugster replied:
Thanks!

joneugster commented on Mathlib/Tactic/FunProp/Theorems.lean:
```diff
@@ -18,104 +18,74 @@ open Lean Meta
+  | const
+  /-- Apply theorem e.g. `Continuous fun (f : (x : X) → Y x => f x)` -/
+  | apply
+  /-- Composition theorem e.g.  `Continuous f → Continuous g → Continuous fun x => f (g x)`
```
```suggestion
  /-- Composition theorem e.g. `Continuous f → Continuous g → Continuous fun x => f (g x)`
```

joneugster commented on Mathlib/Tactic/FunProp/Theorems.lean:
```diff
@@ -18,104 +18,74 @@ open Lean Meta
+  | apply
+  /-- Composition theorem e.g.  `Continuous f → Continuous g → Continuous fun x => f (g x)`
+
+  The numbers `fArgId` and `gArgId` store the arguement index for `f` and `g` in the composition
```
```suggestion
  The numbers `fArgId` and `gArgId` store the argument index for `f` and `g` in the composition
```

joneugster commented on Mathlib/Tactic/FunProp/Theorems.lean:
```diff
@@ -18,104 +18,74 @@ open Lean Meta
+  | const
+  /-- Apply theorem e.g. `Continuous fun (f : (x : X) → Y x => f x)` -/
+  | apply
+  /-- Composition theorem e.g.  `Continuous f → Continuous g → Continuous fun x => f (g x)` -/
```
```suggestion
  /-- Composition theorem e.g. `Continuous f → Continuous g → Continuous fun x => f (g x)` -/
```

joneugster commented on Mathlib/Tactic/FunProp/Theorems.lean:
```diff
@@ -18,104 +18,74 @@ open Lean Meta
 
-set_option linter.unusedVariables false in
-/--  -/
+-- set_option linter.unusedVariables false in
```
```suggestion
```

joneugster commented on Mathlib/Tactic/FunProp/Elab.lean:
```diff
@@ -30,47 +33,60 @@ private def emptyDischarge : Expr → MetaM (Option Expr) :=
+        msg := msg ++ "Issues:"
+        msg := s.mainMsgLog.foldl (init := msg) (fun msg m => msg ++ "\n  " ++ m)
+
+        -- todo enable only when an option is set
```
TODO. See comment about logging above.

joneugster commented on Mathlib/Tactic/FunProp/Elab.lean:
```diff
@@ -17,9 +17,12 @@ namespace Meta.FunProp
 open Lean.Parser.Tactic
 
+/-- `fun_prop` config elaborator -/
+declare_config_elab elabFunPropConfig   FunProp.Config
```
```suggestion
declare_config_elab elabFunPropConfig FunProp.Config
```

joneugster reviewed (changes requested):
I'll stop my review here for now, mainly addressing the logging question. I'll look at the rest later.

Generally, it does help a lot if PRs containing several, mostly independent, changes are split into a few smaller PRs containing a single self-contained feature. Unfortunately, reviewing time grows faster than linear with the PR size.

Further, I already marked various styling issues (like double whitespace), but maybe you could just have another cleanup round yourself first.

joneugster commented on Mathlib/Tactic/FunProp/Theorems.lean:
```diff
@@ -33,7 +33,7 @@ is represented by
+  | const
   | proj (x Y : Nat)
   | projDep (x Y : Nat)
   | comp (f g : Nat)
```
believe this has been addressed.

lecopivo commented:
> I'll stop my review here for now, mainly addressing the logging question. I'll look at the rest later.
> 
> Generally, it does help a lot if PRs containing several, mostly independent, changes are split into a few smaller PRs containing a single self-contained feature. Unfortunately, reviewing time grows faster than linear with the PR size.
> 
> Further, I already marked various styling issues (like double whitespace), but maybe you could just have another cleanup round yourself first.

Sorry about this. This PR was supposed to be just a fix to few bugs but then it got blocked because there was tons of missing docs. I got a bit frustrated by it as I didn't have the time to write it back then. As it is now bug fixes+doc PR I thought I can also squeeze update to logging.

Anyway, thanks a lot for picking on logging as I definitely left it half unfinished. I will clean it up now!

lecopivo commented:
@joneugster Thanks you very much for a great review! Hopefully I have addressed all your comments.

I added a new option `maxTransitionDepth` and `transitionDepth` context variable which limits how many transition theorems can be used. This allowed for few simplification. Logging happens only with `transitionDepth=0` and `thrmStack` has been completely removed which was an incomplete solution to prevent loops anyway.

joneugster reviewed (approved):
Looks good to me! Once all remaining conversations are resolved, I'll put this on the maintainer-merge queue

joneugster commented on Mathlib/Tactic/FunProp/Core.lean:
```diff
@@ -160,77 +160,105 @@ def tryTheorem? (e : Expr) (thmOrigin : Origin) (funProp : Expr → FunPropM (Op
+  for thm in thms do
+    let .id id_X := thm.thmArgs | return none
+    if let .some r ← tryTheoremWithHint? e (.decl thm.thmName) #[(id_X,X)] funProp then
+      return r
```
Looks like the marked sections have docstrings and are quite self-explanatory now. Going to resolve this.

joneugster commented on Mathlib/Tactic/FunProp/Core.lean:
```diff
@@ -160,77 +160,105 @@ def tryTheorem? (e : Expr) (thmOrigin : Origin) (funProp : Expr → FunPropM (Op
 
 /-- Apply lambda calculus rule P fun f => f i` -/
 def applyProjRule (funPropDecl : FunPropDecl) (e x XY : Expr)
     (funProp : Expr → FunPropM (Option Result)) : FunPropM (Option Result) := do
```
Looks like this is resolved.

joneugster commented on test/fun_prop_dev.lean:
```diff
@@ -44,11 +44,12 @@ variable [Obj α] [Obj β] [Obj γ] [Obj δ] [∀ x, Obj (E x)]
 @[fun_prop] theorem Con_applyDep (x : α) : Con (fun f : (x' : α) → E x' => f x) := silentSorry
 @[fun_prop] theorem Con_comp (f : β → γ) (g : α → β) (hf : Con f) (hg : Con g) : Con (fun x => f (g x)) := silentSorry
-@[fun_prop] theorem Con_let (f : α → β → γ) (g : α → β) (hf : Con (fun (x,y) => f x y)) (hg : Con g) : Con (fun x => let y:= g x; f x y) := silentSorry
+-- @[fun_prop] theorem Con_let (f : α → β → γ) (g : α → β) (hf : Con (fun (x,y) => f x y)) (hg : Con g) : Con (fun x => let y:= g x; f x y) := silentSorry
```
support for functions of the form `fun x => let y:= g x; f x y` can be a new PR.

lecopivo commented:
I think I have resolved all the remaining issues.

joneugster commented on Mathlib/Tactic/FunProp/Theorems.lean:
Main change is that the inductive types have been simplified, along with other simplifications.

joneugster commented on Mathlib/Tactic/FunProp/Elab.lean:
This is mostly refactoring code, plus a small fix about dealing with `let`.

joneugster commented on Mathlib/Tactic/FunProp/Core.lean:
This has a series of clean-ups to all the `applyXXXRule` declarations.

joneugster reviewed (approved):
maintainer merge

This is quite a big PR, so I made an experiment and left a comment with a summary on each file with a bigger modification.

dupuisf commented:
bors r+
